vendor/bat-native-ads/src/bat/ads/internal/ads_impl.cc

```diff
@@ -553,6 +565,100 @@ AdsHistory AdsImpl::GetAdsHistory(
   return ads_history;
 }
 
+void AdsImpl::GetPublisherAds(
+    const std::string& url,
+    const std::vector<std::string>& sizes,
+    GetPublisherAdsCallback callback) {
+  if (!ads_client_->ShouldShowPublisherAdsOnParticipatingSites()) {
+    BLOG(INFO) << "Publisher ads are disabled";
+    return;
+  }
+
+  BLOG(INFO) << "Getting publisher ads:"
+      << std::endl << "  URL: " << url
+      << std::endl << "  sizes:" << base::JoinString(sizes, ", ");
+
+  std::vector<std::string> categories = GetWinningCategories();
+  categories.push_back(kUntargetedPageClassification);
+
+  ads_client_->GetCreativePublisherAds(url, categories, sizes,
+      std::bind(&AdsImpl::OnGetCreativePublisherAds, this, callback,
+          _1, _2, _3, _4, _5));
+}
+
+void AdsImpl::OnGetCreativePublisherAds(
+    GetPublisherAdsCallback callback,
+    const Result result,
+    const std::string& url,
+    const std::vector<std::string>& categories,
+    const std::vector<std::string>& sizes,
+    const CreativePublisherAdList& creative_publisher_ads) {
+  PublisherAds ads;
+
+  if (creative_publisher_ads.empty()) {
+    BLOG(INFO) << "No creative publisher ads found for URL: " << url
+        << std::endl << "  categories: " << base::JoinString(categories, ", ")
+        << std::endl << "  sizes: " << base::JoinString(sizes, ", ");
+
+    callback(result, url, sizes, ads);
+    return;
+  }
+
+  auto eligible_creative_publisher_ads =
+      GetEligibleCreativePublisherAds(creative_publisher_ads);
+  if (eligible_creative_publisher_ads.empty()) {
+    BLOG(INFO) << "No eligible publisher ads found for URL: " << url
+        << std::endl << "  categories: " << base::JoinString(categories, ", ")
+        << std::endl << "  sizes: " << base::JoinString(sizes, ", ");
+
+    callback(result, url, sizes, ads);
+    return;
+  }
+
+  BLOG(INFO) << "Found " << eligible_creative_publisher_ads.size()
+      << " eligible creative publisher ads";
+
+  for (const auto& creative_publisher_ad : eligible_creative_publisher_ads) {
+    PublisherAdInfo ad;
+    ad.creative_instance_id = creative_publisher_ad.creative_instance_id;
+    ad.creative_set_id = creative_publisher_ad.creative_set_id;
+    ad.category = creative_publisher_ad.category;
+    ad.size = creative_publisher_ad.size;
+    ad.creative_url = creative_publisher_ad.creative_url;
+    ad.target_url = creative_publisher_ad.target_url;
+    ads.entries.push_back(ad);
+
+    BLOG(INFO) << "Publisher ad for URL: " << url
+        << std::endl << "  creativeInstanceId: " << ad.creative_instance_id
```

Comment:
`std::endl` flushes the buffer every time.  Use `\n` instead.

Reply:
@gdregalo: Legacy so need to change in many places, will raise a new issue for this